### PR TITLE
fixes #2539 - puppetrun API added.

### PR DESCRIPTION
**Pull request by isratrade**
this pr replaces @arnoudj's two commit from pr #636
closing pr #636

### CODE REVIEW

File: app/controllers/api/v2/hosts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module Api
+  module V2
+    class HostsController < V1::HostsController
+      include Api::Version2
+      include Api::TaxonomyScope
```
Comment (ohadlevy) [inline app/controllers/api/v2/hosts_controller.rb]: do we really need the taxonomy scope?

**Comment by ohadlevy** [comment]
note that ticket number is wrong, it should be 2539

File: app/controllers/api/v2/hosts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+      before_filter :find_resource, :only => %w{show update destroy status puppetrun}
```
Comment (ohadlevy) [inline app/controllers/api/v2/hosts_controller.rb]: dont we have before filter on the parent class? do we need to define it twice?

**Comment by isratrade** [comment]
commit messages updated

**Comment by isratrade** [comment]
recommited

File: app/controllers/api/v2/hosts_controller.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+module Api
+  module V2
+    class HostsController < V1::HostsController
+      include Api::Version2
+
+      api :GET, "/hosts/:id/puppetrun", "Force a puppet run on the agent."
+
+      def puppetrun
+        process_response @host.puppetrun!
```
Comment (ohadlevy) [inline app/controllers/api/v2/hosts_controller.rb]: what sets @host here then?

Comment (isratrade) [inline app/controllers/api/v2/hosts_controller.rb]: recommitted with before_filter added back

File: lib/foreman/access_permissions.rb [hunk 4]
```diff
@@ -234,7 +234,7 @@
                                     :puppetclasses => pc_ajax_actions,
                                     :subnets => subnets_ajax_actions,
                                     :"api/v1/hosts" => [:update],
-                                    :"api/v2/hosts" => [:update]
+                                    :"api/v2/hosts" => [:update, :puppetrun]
```
Comment (domcleal) [inline lib/foreman/access_permissions.rb]: This should be added to `puppetrun_hosts` instead.

**Comment by isratrade** [comment]
@domcleal, re-commit ed with your fix

File: app/controllers/api/v2/hosts_controller.rb (hunk 3 above)
Comment (domcleal) [inline app/controllers/api/v2/hosts_controller.rb]: The main hosts controller checks Setting[:puppetrun] here and denies it if disabled, we should probably copy that here.

**Comment by isratrade** [comment]
@domcleal, good catch.  re-committed with check for Setting[:puppetrun]

**Comment by domcleal** [comment]
@isratrade tests are failing now, you might need to change the puppetrun setting now to true in the test, before calling the API?

**Comment by isratrade** [comment]
@domcleal, re-committed.  good catch.  I'm still haven't problems running my local tests .... otherwise I could catch it before pushing.

**Comment by domcleal** [comment]
@isratrade tests failed again :(

I suspect you need to add the puppetrun setting into the test fixtures.

**Comment by ohadlevy** [comment]
merged at f2fed49 and 0fda4cf - thanks!

**Pull request closed without merging**